extract google pubsub

dvoet opened the pull request:
https://broadinstitute.atlassian.net/browse/GAWB-1653

The primary purpose of this PR is to pull the google pubsub code into a place that can then be used by thurloe. This is done by following the pattern set for the model library. Two new libraries have been created: google and util. Next step is to either publish these to artifactory or pull them into new repos and use google submodules. I am going to explore the former first.

- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [x] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [x] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
- [x] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [x] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [x] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [x] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [x] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

dvoet committed:
wip

dvoet committed:
right dependency

dvoet committed:
wip

dvoet committed:
wip

dvoet committed:
tests passing

dvoet committed:
error report source

dvoet committed:
move Retry to util

dvoet committed:
moved google code to google package

dvoet committed:
moved ErrorReportable to model

dvoet committed:
exclude guava

dvoet commented on model/src/main/scala/org/broadinstitute/dsde/rawls/model/ErrorReportable.scala:
```diff
@@ -1,7 +1,6 @@
-package org.broadinstitute.dsde.rawls.dataaccess
+package org.broadinstitute.dsde.rawls.model
```
I really feel like all the ErrorReport stuff goes in util but not ready to do that yet. That will come when PerRequest is moved to util.

helgridly replied:
Agreed.

helgridly commented on build.sbt:
```diff
@@ -2,17 +2,30 @@ import Settings._
 import Testing._
 
 
+lazy val rawlsUtil = project.in(file("util"))
```
do we want to think about not calling these rawlsFoo in favour of workbenchFoo or something?

helgridly commented on build.sbt:
```diff
@@ -2,17 +2,30 @@ import Settings._
 
+lazy val rawlsGoogle = project.in(file("google"))
+  .settings(googleSettings:_*)
+  .dependsOn(rawlsModel)
```
Really? That's a shame.

dvoet replied:
because of ErrorReport stuff which will get moved to util soon I hope

helgridly commented on build.sbt:
```diff
@@ -2,17 +2,30 @@ import Settings._
 
 lazy val rawls = project.in(file("."))
   .settings(rootSettings:_*)
+  .aggregate(rawlsGoogle)
```
Definitely run and test assembly here. The difference between aggregate and dependsOn tripped me up a lot.

helgridly commented on google/src/main/scala/org/broadinstitute/dsde/rawls/google/GoogleUtilities.scala:
```diff
@@ -94,8 +94,8 @@ trait GoogleUtilities extends LazyLogging with Retry {
   private def logGoogleRequest[A](request: AbstractGoogleClientRequest[A], startTime: Long, statusCode: Option[Int], errorReport: Option[ErrorReport]): Unit = {
-    import spray.json._
     import GoogleRequestJsonSupport._
+    import spray.json._
```
In-function imports shouldn't really be necessary any more.

dvoet replied:
since GoogleRequestJsonSupport is a protected member, it can't go at the top so I think I will let this ride... in this particular case it is only in the function because it is only used here

helgridly commented on google/src/main/scala/org/broadinstitute/dsde/rawls/google/package.scala:
```diff
@@ -0,0 +1,7 @@
+
+package object google {
+  implicit val errorReportSource = ErrorReportSource("google")
+}
```
![](https://upload.wikimedia.org/wikipedia/en/f/ff/SuccessKid.jpg)

helgridly commented on model/src/main/scala/org/broadinstitute/dsde/rawls/model/ErrorReportable.scala:
```diff
@@ -12,11 +11,11 @@ trait ErrorReportable {
       case gjre: GoogleJsonResponseException =>
         val statusCode = StatusCodes.getForKey(gjre.getStatusCode)
-        ErrorReport(ErrorReport.message(gjre), statusCode, ErrorReport.causes(gjre), Seq.empty, Option(gjre.getClass))(errorReportSource)
+        ErrorReport(ErrorReport.message(gjre), statusCode, ErrorReport.causes(gjre)(errorReportSource), Seq.empty, Option(gjre.getClass))(errorReportSource)
```
Shouldn't this be implicit-y?

dvoet replied:
I thought the whole point of this class was to define a source other than the one defined implicit-y. Before the code was moved it was using the local errorReportSource for the top ErrorReport but the implicit source for the causes. So, it might look like the top is from "google" and the causes are from "rawls." This change makes the whole stack come from "google" (or whatever errorReportSource is set to be).

helgridly replied:
Hmm. That being the case, can you update [this test](https://github.com/broadinstitute/rawls/blob/develop/core/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionSpec.scala#L546) with whatever's appropriate?

dvoet replied:
I would need a special test for this, that test case does not have an exception that lists causes

helgridly reviewed (approved):
no blocking changes, but some questions and nitpicks

dvoet committed:
aggregate util too

dvoet committed:
added test

dvoet committed:
workbench prefix for new artifacts

dvoet committed:
update publishing

dvoet commented:
@helgridly I added a test to cover an error report with a cause that has a root cause. I also changed where we publish to org.broadinstitute.dsde (I added dsde) and I changed the rawls-util and rawls-google artifacts to be workbench-util and workbench-google. And I fiddled with the publish scripts.

dvoet committed:
centralize toScalaDuration

helgridly commented:
👍 please update orch to refer to the relocated rawlsModel once you've merged this
